### ISSUE TITLE
How do you deal with multiple peaks in final eval scores?

### ISSUE THREAD

**Issue by FrancescoCappio** [hongjoon0805/SS-IL-Official#8]
Hello!
I really like the principle behind your idea of separated softmax, however while trying to implement it in my class incremental framework I noticed **a possible downside** of your approach: it encourages the model to produce **multiple peaks** in the output scores, with the peak for the correct class not being necessarily the highest one. 
Let me explain with one example: 
we are trying to train SS-IL in an incremental setting with just 2 tasks. During the first we learn classes from 0 to 9, during the second those from 10 to 19. 
At training time, besides the loss, we compute also the **classification accuracy on train data**. When training on the second task we can compute 3 different accuracies:

- accuracy on **old** data. We select old data (from replay buffer) with corresponding ground truth labels, look at the output scores for old classes (those of the **previous** task), select the predicted class using the index of the highest score and compute the accuracy by comparing ground truth labels and predicted classes;
- accuracy on **new** data. We select new data (from training batch) with corresponding ground truth labels, look at the output scores for new classes (those of the **current** task), select the predicted class using the index of the highest score and compute the accuracy by comparing ground truth labels and predicted classes;
- accuracy on **all** data. We select old data (from replay buffer) and new data (from training batch) with corresponding ground truth labels. We use the **complete output scores** to see which is the predicted class (the one with highest score) and compute the accuracy by comparing ground truth labels and predicted classes.

As you can see, for the first two metrics we apply **the same separation criterion** that you propose for the softmax on the training loss. We can do so because **at training time we do know the ground truth label** of each sample and therefore we can separate the scores for the two tasks. 
For what concerns the last metric it is more like the accuracy that you compute at inference time when the test sample could belong to classes of any of the learned tasks. 

In my experiments I have **very high values for the first two metrics**, while having **a quite low value for the last one**. This can be explained easily: **prediction for a certain sample is correct when considering only the scores for the classes of its own task**, but when we look at the whole score vector **there may be an higher value among the scores produced for classes of the other task**. This is a direct consequence of the fact that the separated softmax strategy does not push for the lowering of scores produced for old classes. In practice the separated softmax strategy aims at maximizing the first two metrics, while it does not directly push for the maximization of the third. So we can say that **the learning strategy is working correctly** (I have high values for the first two metrics) but **it is not doing what we really want** (i.e. to maximize the accuracy on all data). 

The consequence of this behavior is that at the end I get very bad overall performance from SS-IL. What's your opinion on this problem? How did you manage to solve it in your implementation in order to obtain the good results that you report in your paper?

**Comment by renyixin666** [comment]
I have the same concern, will the details on implementation solve this problem?

**Comment by hongjoon0805** [comment]
Hi!
Sorry for the late reply

We agree with your concern. Because of the lack of comparison between two peak points on old and new classes during training, our method do not have the ability to separate those classes.

The reason for achieving high accuracy in large scale incremental learning is that the root cause of the performance degradation is prediction bias, and separating the softmax during training decreases the score bias which mainly contributes to the prediction bias.

In small scale incremental learning, we empirically find that the prediction bias does not contribute to the large portion on performance degradation, and those kind of results are already proposed in [Wu et. al.].

Therefore, if you try to achieve high accuracy using SS-IL in small scale incremental learning setting (e.g. CIFAR-100 or mini-Imagenet), it does not work well. 

If you have trouble in reproducing our results on Imagenet-1K or Google Landmark-v2 10K dataset, please let us know.
